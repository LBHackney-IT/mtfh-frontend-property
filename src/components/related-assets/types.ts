export interface RelatedAsset {
  id: string;
  name: string;
  type: string;
}
