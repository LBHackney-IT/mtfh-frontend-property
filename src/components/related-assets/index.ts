export * from "./related-asset";
export * from "./related-assets";
export * from "./related-asset-group";
