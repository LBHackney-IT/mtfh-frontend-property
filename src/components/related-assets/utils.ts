import { sortAddressGeneric } from "../../utils/sortAddressGeneric";
import { RelatedAsset } from "./types";

import { Asset, ParentAsset } from "@mtfh/common/lib/api/asset/v1";

// This methods outputs a unique list of related assets, all in the same format (RelatedAsset interface)
export const getAllRelatedAssets = (
  parentAssets: ParentAsset[],
  childrenAssets: Asset[] = [],
) => {
  const allRelatedAssets: RelatedAsset[] = parentAssets;

  if (childrenAssets?.length) {
    const formattedChildrenAssets = childrenAssets.map((childrenAsset) => ({
      type: childrenAsset.assetType,
      id: childrenAsset.id,
      name: childrenAsset.assetAddress.addressLine1,
    }));

    allRelatedAssets.push(...formattedChildrenAssets);
  }

  return allRelatedAssets;
};

export const organiseRelatedAssetsByType = (
  relatedAssets: RelatedAsset[],
): { [key: string]: RelatedAsset[] } => {
  const assetsByType: { [key: string]: RelatedAsset[] } = {};

  // Define how many asset types we're dealing with
  const uniqueAssetTypes = new Set<string>(relatedAssets.map((x) => x.type));

  // If present, move "Dwelling" to last (so dwellings appear below other asset types)
  if (uniqueAssetTypes.has("Dwelling")) {
    uniqueAssetTypes.delete("Dwelling");
    uniqueAssetTypes.add("Dwelling");
  }

  // For each AssetType, create a array of RelatedAsset[]
  uniqueAssetTypes.forEach((uniqueAssetType) => {
    const sameTypeAssets: RelatedAsset[] = relatedAssets.filter(
      (x) => x.type === uniqueAssetType,
    );

    sortAddressGeneric(sameTypeAssets, "name");

    // Create new key in object for given AssetType, value will be an array related assets of that type
    assetsByType[uniqueAssetType] = sameTypeAssets;
  });

  return assetsByType;
};
