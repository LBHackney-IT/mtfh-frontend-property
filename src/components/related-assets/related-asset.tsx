import React from "react";
import { Link as RouterLink } from "react-router-dom";

import { SearchCard } from "../../search/components/search-card";
import { RelatedAsset as IRelatedAsset } from "./utils";

import { Link, LinkBox, LinkOverlay } from "@mtfh/common/lib/components";

export interface RelatedAssetProps {
  relatedAsset: IRelatedAsset;
}

export const RelatedAsset = ({ relatedAsset }: RelatedAssetProps) => {
  return (
    <LinkBox data-testid="related-asset">
      <SearchCard>
        <LinkOverlay>
          <Link
            className="mtfh-search-tenure__title"
            as={RouterLink}
            to={`/property/${relatedAsset.id}`}
            variant="text-colour"
          >
            {relatedAsset.name}
          </Link>
        </LinkOverlay>
      </SearchCard>
    </LinkBox>
  );
};
