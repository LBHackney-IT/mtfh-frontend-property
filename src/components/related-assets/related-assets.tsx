import React from "react";
import { Link as RouterLink } from "react-router-dom";

import { Link, LinkBox, LinkOverlay } from "@mtfh/common/lib/components";
import { SearchCard } from "@mtfh/search";

import { RelatedAsset as IRelatedAsset } from "views/related-assets-view/utils";

export interface RelatedAssetsProps {
  assetType: string;
  relatedAssets: IRelatedAsset[];
}

export const RelatedAssets = ({ assetType, relatedAssets }: RelatedAssetsProps) => {
  const getAssetTypeHeading = () => {
    if (assetType.charAt(assetType.length - 1).toLowerCase() === "s") {
      return assetType;
    }
    if (assetType === "NA") {
      return "N/A";
    }
    return `${assetType}s`;
  };

  const renderRelatedAssets = () => {
    return relatedAssets.map((relatedAsset) => {
      return <RelatedAsset relatedAsset={relatedAsset} key={relatedAsset.id} />;
    });
  };

  return (
    <section>
      <p className="lbh-body-m">{getAssetTypeHeading()}</p>
      <div className="mtfh-card-list" data-testid="related-asset-items-group">
        {renderRelatedAssets()}
      </div>
    </section>
  );
};

export interface RelatedAssetProps {
  relatedAsset: IRelatedAsset;
}

export const RelatedAsset = ({ relatedAsset }: RelatedAssetProps) => {
  return (
    <LinkBox data-testid="related-asset">
      <SearchCard>
        <LinkOverlay>
          <Link
            className="mtfh-search-tenure__title"
            as={RouterLink}
            to={`/property/${relatedAsset.id}`}
            variant="text-colour"
          >
            {relatedAsset.name}
          </Link>
        </LinkOverlay>
      </SearchCard>
    </LinkBox>
  );
};
