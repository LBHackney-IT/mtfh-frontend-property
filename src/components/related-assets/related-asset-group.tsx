import React from "react";

import { RelatedAsset } from "./related-asset";
import { RelatedAsset as IRelatedAsset } from "./types";

interface RelatedAssetGroupProps {
  assetType: string;
  relatedAssets: IRelatedAsset[];
}

export const RelatedAssetGroup = (props: RelatedAssetGroupProps) => {
  const { assetType, relatedAssets } = props;

  const headingIsPlural = assetType.charAt(assetType.length - 1).toLowerCase() === "s";

  return (
    <section>
      <p className="lbh-body-m">{headingIsPlural ? assetType : `${assetType}s`}</p>
      <div className="mtfh-card-list" data-testid="related-asset-items-group">
        <>
          {relatedAssets.map((relatedAsset) => (
            <RelatedAsset relatedAsset={relatedAsset} key={relatedAsset.id} />
          ))}
        </>
      </div>
    </section>
  );
};
