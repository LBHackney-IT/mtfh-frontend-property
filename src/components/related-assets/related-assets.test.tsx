import React from "react";

import { render } from "@hackney/mtfh-test-utils";

import { RelatedAssets } from "./related-assets";

const address1 = "123 Test Street";
const address2 = "456 Test Road";
const address3 = "789 Test Estate";

const assetsByType = {
  Estate: [
    {
      type: "Estate",
      id: "14edf718-19ff-ff43-4679-f8ef404fa029",
      name: address3,
    },
  ],
  Dwelling: [
    {
      type: "Dwelling",
      id: "13e7cf17-60a0-729d-c297-a4e76a16b6fa",
      name: address1,
    },
    {
      type: "Dwelling",
      id: "ea0b9c3e-3d72-276c-be2d-04620e2bf0e3",
      name: address2,
    },
  ],
};

test("it renders the component", async () => {
  const assetType = "Dwelling";
  const { container } = render(
    <RelatedAssets assetType={assetType} relatedAssets={assetsByType[assetType]} />,
  );
  expect(container).toMatchSnapshot();
});

test("the asset type is displayed in plural form", async () => {
  const assetType = "Dwelling";

  const { container } = render(
    <RelatedAssets assetType={assetType} relatedAssets={assetsByType[assetType]} />,
  );
  expect(container).toHaveTextContent(`${assetType}s`);
});

test("related assets' addresses are displayed", async () => {
  const assetType = "Dwelling";

  const { container } = render(
    <RelatedAssets assetType={assetType} relatedAssets={assetsByType[assetType]} />,
  );

  expect(container).toHaveTextContent(address1);
  expect(container).toHaveTextContent(address2);
});
