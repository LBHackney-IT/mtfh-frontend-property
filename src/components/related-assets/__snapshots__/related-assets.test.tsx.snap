// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders the component 1`] = `
<div>
  <section
    data-testid="related-assets-by-type"
    id="related-assets-by-type"
  >
    <p
      class="lbh-body-m"
    >
      Dwellings
    </p>
    <div
      class="mtfh-card-list"
      data-testid="related-asset-items"
    >
      <div
        class="mtfh-link-box"
      >
        <div
          class="mtfh-search-card"
        >
          <div
            class="mtfh-link-overlay"
          >
            <a
              class="govuk-link lbh-link lbh-link--text-colour lbh-link--no-visited-state mtfh-search-tenure__title"
              href="/property/13e7cf17-60a0-729d-c297-a4e76a16b6fa"
            >
              123 Test Street
            </a>
          </div>
        </div>
      </div>
      <div
        class="mtfh-link-box"
      >
        <div
          class="mtfh-search-card"
        >
          <div
            class="mtfh-link-overlay"
          >
            <a
              class="govuk-link lbh-link lbh-link--text-colour lbh-link--no-visited-state mtfh-search-tenure__title"
              href="/property/ea0b9c3e-3d72-276c-be2d-04620e2bf0e3"
            >
              456 Test Road
            </a>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
